pkg/provisioner/ironic/ironic.go

```diff
@@ -1346,3 +1356,51 @@ func (p *ironicProvisioner) PowerOff() (result provisioner.Result, err error) {
 
 	return result, nil
 }
+
+// SoftPowerOff tries to perform soft shutdown first
+// If soft shutdown is unspported, it performs hard shutdown
+func (p *ironicProvisioner) SoftPowerOff() (result provisioner.Result, err error) {
+	p.log.Info("ensuring host is soft powered off")
+
+	ironicNode, err := p.findExistingHost()
+	if err != nil {
+		return result, errors.Wrap(err, "failed to find existing host")
+	}
+
+	if ironicNode.PowerState != powerOff {
```

Comment:
See if you can reorganize the logic of this function to include more early exits and fewer levels of indentation. For example, here, if ironicNode.PowerState == powerOff { ... return } would mean we would not need an "else" block, and all of the code on lines 1370-1399 could be dedented one level.